pallets/msa/src/lib.rs

```diff
@@ -1292,6 +1360,87 @@ impl<T: Config> Pallet<T> {
 		block_number / (T::BlockNumber::from(T::MortalityWindowSize::get())) %
 			T::BlockNumber::from(T::NumberOfBuckets::get())
 	}
+
+	/// Get list of account ids that are associated with a given msa
+	pub fn get_msa_keys_offchain(msa_id: MessageSourceId) -> (T::BlockNumber, Vec<T::AccountId>) {
+		let msa_keys_res = get_msa_keys::<MessageSourceId, T::AccountId, T::BlockNumber>(msa_id);
+		if let Err(e) = msa_keys_res {
+			log_err!("Error getting msa keys from offchain storage: {:?}", e);
+			return (T::BlockNumber::zero(), vec![])
+		}
+		msa_keys_res.unwrap()
+	}
+
+	fn process_events(
+		msa_id: MessageSourceId,
+		events: Vec<(T::AccountId, EventType, T::BlockNumber)>,
+	) {
+		let lock_status = offchain_common::lock(msa_id.encode().as_slice(), || {
+			for (key, event_type, block) in &events {
+				match event_type {
+					EventType::Add => {
+						let add_result =
+							process_msa_key_event::<MessageSourceId, T::AccountId, T::BlockNumber>(
+								MSAPublicKeyDataOperation::Add::<
+									MessageSourceId,
+									T::AccountId,
+									T::BlockNumber,
+								>(msa_id, key.clone(), block.clone()),
+							);
+						if let Err(e) = add_result {
+							log_err!("Error adding key to offchain storage: {:?}", e);
+						}
+					},
+					EventType::Remove => {
+						let delete_result =
+							process_msa_key_event::<MessageSourceId, T::AccountId, T::BlockNumber>(
+								MSAPublicKeyDataOperation::Remove::<
+									MessageSourceId,
+									T::AccountId,
+									T::BlockNumber,
+								>(msa_id, key.clone(), block.clone()),
+							);
+						if let Err(e) = delete_result {
+							log_err!("Error deleting key from offchain storage: {:?}", e);
+						}
+					},
+				}
+			}
+		});
+		if let offchain_common::LockStatus::Locked = lock_status {
+			log_err!("Unable to acquire lock, skipping event processing");
+		}
+	}
+
+	fn index_event(event: Event<T>) {
+		let block_number: u32 =
+			<frame_system::Pallet<T>>::block_number().try_into().unwrap_or_default();
+		let current_event_count: u16 = <MSAEventCount<T>>::get().saturating_add(1);
+		<MSAEventCount<T>>::put(current_event_count);
+		let key = [
+			BLOCK_EVENT_KEY,
+			block_number.encode().as_slice(),
+			current_event_count.encode().as_slice(),
+		]
+		.concat();
+		// set the event in offchain storage
+		set_offchain_index(key.encode().as_slice(), event);
+	}
+
+	fn read_events(block_number: u32, event_count: u16) -> Vec<Event<T>> {
+		let mut events = vec![];
+		for i in 1..event_count + 1 {
+			let key =
+				[BLOCK_EVENT_KEY, block_number.encode().as_slice(), i.encode().as_slice()].concat();
```

Comment:
Same key we set while indexing

Reply:
in theory, you don't need to have the whole key, you should be able to lookup all the keys that have the same prefix, in our case `[BLOCK_EVENT_KEY, block_number.encode().as_slice()]`. So you might not even need the  event_count.

NVM I see that they didn't expose getting values by prefix from offchain storage api which sucks, so it seems like it would be difficult to do that.